### ISSUE TITLE
Checkpoint and fft don't play well togethet after jax > 0.3.20

### ISSUE THREAD

**Issue by astanziola** [google/jax#14302]
### Description

Hello Jax Team,

I have noticed that after upgrading to Jax greater than 0.3.20, the output of my acoustic simulator (based on Jax) had some edge cases where the gradient calculations were incorrect. After some investigation, I believe the issue is related to the combination of the `fftn` functions and `jax.checkpoint` and it happened somewhere between the release of `jax 0.3.21` and `jax 0.3.24`.

I have provided a piece of code that reproduces the issue below. I apologize for its length, but I wasn't able to make it shorter. 

I have also added comments in the code that begin with `### ERROR SOURCE` to highlight some possible changes that make the issue disappear.

```python
from typing import Iterable, NamedTuple

import jax
import numpy as np
from jax import numpy as jnp
from jax.scipy.sparse.linalg import gmres
from scipy.io import savemat


class Domain(NamedTuple):
    N: Iterable[int] = (32, 32)
    dx: Iterable[float] = (1.0, 1.0)

def helmholtz_solver(medium, omega, source):
    helm_func = lambda u: helmholtz(u, medium, omega=omega)

    ### ERROR SOURCE 1
    ### Without the checkpoint decorator, the result is the same across jax versions
    ### comment out the next line to make the problem go away
    helm_func = jax.checkpoint(helm_func)

    out = gmres(
        helm_func,
        source,
        source * 0,
        tol=1e-3,
        restart=10,
        maxiter=1000,
        solve_method="batched",
    )[0]
    return -1j * omega * out


def helmholtz(u, medium, *, omega=1.0):
    L = laplacian_term(u, medium)
    k = ((omega / medium["sound_speed"]) ** 2) * u
    return L + k


def _get_ffts(x, domain):
    if jnp.iscomplexobj(x):
        ffts = [jnp.fft.fft, jnp.fft.ifft]
        _f = lambda N, dx: jnp.fft.fftfreq(N, dx) * 2 * jnp.pi
    else:
        ffts = [jnp.fft.rfft, jnp.fft.irfft]
        _f = lambda N, dx: jnp.fft.rfftfreq(N, dx) * 2 * jnp.pi

    k_axis = [_f(n, delta) for n, delta in zip(domain.N, domain.dx)]
    return ffts, k_axis


def fourier_filter(u, domain, kind="gradient"):
    # Evaluates both the gradient of a scalar function, or the diagonal of the
    # jacobian of a vector function.

    # Make k-vector
    ffts, k_vec = _get_ffts(u, domain)

    # Builds filters
    if kind == "gradient":
        # Filter for directional gradients
        k_vec = [1j * k for k in k_vec]
    elif kind == "identity":
        # A constant filter, equivalent to the identity operator
        k_vec = [k * 0.0 + 1.0 + 0j for k in k_vec]

    def single_grad(axis, u):
        u = jnp.moveaxis(u, axis, -1)
        Fx = ffts[0](u, axis=-1)
        iku = Fx * k_vec[axis]
        du = ffts[1](iku, axis=-1, n=u.shape[-1])
        return jnp.moveaxis(du, -1, axis)

    if u.shape[-1] == 1:
        # Gradient of scalar function
        new_params = jnp.stack([single_grad(i, u[..., 0]) for i in range(len(domain.N))], axis=-1)
    else:
        # Diagonal of jacobian
        new_params = jnp.stack([single_grad(i, u[..., i]) for i in range(len(domain.N))], axis=-1)

    return new_params

def laplacian_term(u, medium):
    # Making laplacian
    grad_u = fourier_filter(u, medium["domain"], "gradient")
    mod_diag_jacobian = fourier_filter(grad_u, medium["domain"], "gradient")
    nabla_u = jnp.sum(mod_diag_jacobian, axis=-1, keepdims=True)

    # Making the density term
    grad_rho0 = fourier_filter(medium["density"], medium["domain"], "gradient")

    ### ERROR SOURCE 2
    ### If I don't apply the identity filter, I get the same results for different jax versions
    _ru = fourier_filter(grad_u * grad_rho0, medium["domain"], "identity")
    # _ru = grad_u * grad_rho0

    rho_u = jnp.sum(_ru, axis=-1, keepdims=True) / medium["density"]

    return nabla_u - rho_u

if __name__ == "__main__":
    from matplotlib import pyplot as plt
    # Setup simulation
    domain = Domain((59, 59, 29), (1.0, 1.0, 1.0))
    sound_speed = jnp.zeros(domain.N + (1,), dtype=jnp.float32) + 1.0
    source = jnp.zeros(domain.N + (1,), dtype=jnp.complex64).at[15:-15, 15:-15, 10, 0].set(1.0)

    @jax.jit
    @jax.grad
    def loss_func(sound_speed, source):
        density = jnp.zeros(domain.N + (1,), dtype=jnp.float32) + 1000.0
        medium = {"domain": domain, "sound_speed": sound_speed, "density": density}

        total_field = helmholtz_solver(medium, 1.0, source)
        lossval = -jnp.sum(jnp.abs(total_field[29, 29, 21, 0]))
        return lossval

    gradient = loss_func(sound_speed, source)
    plt.imshow(gradient[29,:,:,0])
    plt.colorbar()
    plt.savefig("repro_error.png")
```

I have run the code with two different Jax versions, keeping all the rest the same. Following are the relevant outputs of pip list for each environment.

```bash
(jax_old) astanziola@repo/folder/$ pip list
Package             Version               Editable project location
------------------- --------------------- ----------------------------
...
jax                 0.3.20
jaxlib              0.3.20+cuda11.cudnn82

(jax_new) astanziola@kinsler:~/repos/holab$ pip list
Package             Version              Editable project location
------------------- -------------------- ----------------------------
....
jax                 0.4.1
jaxlib              0.4.1+cuda11.cudnn86
```

When look at the variable `gradient`, by taking the slice `gradient[29,:,:,0]`, there are two very different results for the two Jax versions. I have included images of the results for reference.

![image](https://user-images.githubusercontent.com/3237202/216839656-23210c6c-a6a4-4e0d-9d75-1ffc858c7a8b.png)

Some insights that may be helpful in resolving this issue:
- If the checkpoint function is removed or the calculation of `_ru` is changed, the results for the new Jax version change to match the ones from the old code. However, this does not happen with the old Jax version, where the results are always the same.
- The provided code may not make physical sense, but it is stripped from a larger and more complex acoustic simulator that has been experimentally tested. The old Jax version gives the correct results, while the new Jax version gives physically incorrect results.
- The issue only seems to occur for 3D inputs and works fine for 2D inputs. It may have something to do with the 3D version of `fftn`.

### What jax/jaxlib version are you using?

jax 0.3.20 and jax 0.4.1 (but the problem happened since at least ~jax 0.3.24)

### Which accelerator(s) are you using?

GPU

### Additional system info

Ubuntu, python 3.10

### NVIDIA GPU info

```bash
$ nvidia-smi
Sun Feb  5 19:02:05 2023       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 525.78.01    Driver Version: 525.78.01    CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla P40           Off  | 00000000:04:00.0 Off |                  Off |
| N/A   44C    P0   124W / 250W |  22260MiB / 24576MiB |    100%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  Tesla P40           Off  | 00000000:05:00.0 Off |                  Off |
| N/A   26C    P0    50W / 250W |      0MiB / 24576MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   2  Tesla P40           Off  | 00000000:08:00.0 Off |                  Off |
| N/A   25C    P0    49W / 250W |      0MiB / 24576MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   3  Tesla P40           Off  | 00000000:09:00.0 Off |                  Off |
| N/A   26C    P0    49W / 250W |      0MiB / 24576MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   4  Tesla P40           Off  | 00000000:83:00.0 Off |                  Off |
| N/A   28C    P0    50W / 250W |      0MiB / 24576MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   5  Tesla P40           Off  | 00000000:84:00.0 Off |                  Off |
| N/A   31C    P0    50W / 250W |      0MiB / 24576MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   6  Tesla P40           Off  | 00000000:87:00.0 Off |                  Off |
| N/A   24C    P0    49W / 250W |      0MiB / 24576MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   7  Tesla P40           Off  | 00000000:88:00.0 Off |                  Off |
| N/A   26C    P0    49W / 250W |      0MiB / 24576MiB |      1%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
```

```bash
$ nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2022 NVIDIA Corporation
Built on Tue_May__3_18:49:52_PDT_2022
Cuda compilation tools, release 11.7, V11.7.64
Build cuda_11.7.r11.7/compiler.31294372_0
```

**Comment by astanziola** [comment]
It looks like this is not an issue anymore with jax 0.4.11, I guess this can be closed.

**Comment by hawkinsp** [comment]
Closing!
